packages/components/src/navigator/match-path.js

```diff
@@ -0,0 +1,79 @@
+/**
+ * External dependencies
+ */
+import pathToRegexp from 'path-to-regexp';
+
+const cache = {};
+const cacheLimit = 10000;
+let cacheCount = 0;
+
+function compilePath( path, options ) {
+	const cacheKey = `${ options.end }${ options.strict }${ options.sensitive }`;
+	const pathCache = cache[ cacheKey ] || ( cache[ cacheKey ] = {} );
+
+	if ( pathCache[ path ] ) return pathCache[ path ];
+
+	const keys = [];
+	const regexp = pathToRegexp( path, keys, options );
+	const result = { keys, regexp };
+
+	if ( cacheCount < cacheLimit ) {
+		pathCache[ path ] = result;
+		cacheCount++;
+	}
+
+	return result;
+}
+
+/**
+ * Public API for matching a URL pathname to a path.
+ *
+ * @param {string}       pathname Pathname
+ * @param {string|Array} options  Matching options.
+ * @return {Object} Matched path details
+ */
+function matchPath( pathname, options = {} ) {
```

Comment:
I'm missing some context on this function.
I guess when isExact is true we verify if pathname is equal to options.path if yes we return true. This is the most common case, should we have a special condition to avoid regex, etc on this case and just compare the strings? (it seems a good performance improvement).

When isExact is false what patches do we match? Is it just a path like "blocks/" matches all paths with shape "blocks/.../...". If yes do we need a regex for that? Could we just verify if the string starts with "blocks/"? I guess the issue is much more complex and we accept a diverse set of matches but I'm missing context on what we are supposed to accept and not so I'm having trouble properly reviewing the algorithm.